### PR TITLE
Prepare Dev Services for Keycloak config at run time

### LINKED ISSUE #21047: in version 2.3.1 and 2.4.0 Vault MicroProfile Config Source stopped working with OIDC
### Describe the bug

in version 2.3.1 and 2.4.0 Vault MicroProfile Config Source stopped working to use vault with OIDC (Vault MicroProfile Config Source)

### Expected behavior

quarkus fetched vault kv values first and then try verificted dependent properties. In version 2.1.2 works fine, but in 2.3.1 and 2.4.0 does not work

### Actual behavior

on startup throw error:

```
2021-10-27 15:40:13,605 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:418)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:276)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:67)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:92)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:441)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:65)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.builder.Execution.run(Execution.java:116)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:161)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:416)
	... 9 more
Caused by: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Linux localhost 5.14.11-1-default #1 SMP Sun Oct 10 08:34:34 UTC 2021 (834dddd) x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.12" 2021-07-20 LTS OpenJDK Runtime Environment Corretto-11.0.12.7.1 (build 11.0.12+7-LTS) OpenJDK 64-Bit Server VM Corretto-11.0.12.7.1 (build 11.0.12+7-LTS, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.4.0.FINAL

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Hi @radcortez, @vsevel: please note that this error is similar to old #14707 but only with OIDC.

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #21047

### CODE REVIEW

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java [hunk 1]
```diff
@@ -220,28 +223,38 @@ public void run() {
...
 
-    private KeycloakDevServicesConfigBuildItem prepareConfiguration(boolean createRealm, String realmNameToUse,
-            BuildProducer<DevServicesConfigResultBuildItem> devServices) {
-        final String realmName = realmNameToUse != null ? realmNameToUse : getDefaultRealmName();
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = { IsEnabled.class, GlobalDevServicesConfig.Enabled.class })
+    @Record(ExecutionTime.RUNTIME_INIT)
+    public KeycloakDevServicesConfigBuildItem prepareConfiguration(
+            Optional<KeycloakDevServicesStartedBuildItem> devServicesStarted,
+            BuildProducer<DevServicesConfigResultBuildItem> devServicesConfig,
+            ConfigRecorder recorder) {
```
Comment (gsmet) [inline extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:236]: I don't see the recorder being used here and the non recorded code is executed at build time so I'm not sure what you did there?

**Review by gsmet — DISMISSED** [review]
Let's clarify my question before merging.

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 1 above)
Comment (stuartwdouglas) [reply extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:236]: In that case the method used to record, and when it was changed someone (probably me), missed removing the recorder. I have fixed it in https://github.com/quarkusio/quarkus/pull/21325

The solution is to remove `@Record`, not to add an unused recorder.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
I really don't see how this relates to the linked issue? The linked issue is with the OIDC Dev UI, not DevServices. It needs to be fixed by changing to `DevConsoleRuntimeTemplateInfoBuildItem` to evaluate these config values at runtime.

**Comment by sberyozkin** [comment]
Hi Stuart @stuartwdouglas `KeycloakDevServicesProcessor` checks `quarkus.oidc.client-id` and `quarkus.oidc.credentials.secret` to create a default realm. Without this fix, the added test fails - this means the fix is good ?

**Comment by sberyozkin** [comment]
@stuartwdouglas By the way, the linked issue is #21047, I only linked to #14309 to clarify why I had to add the `@Recorder`

**Comment by sberyozkin** [comment]
@stuartwdouglas Oops, sorry, yeah, there is one more issue to fix

**Comment by sberyozkin** [comment]
@stuartwdouglas @radcortez 

So I've update 2 tests:
 * `integration-tests/oidc-token-propagation` with Dev Services for Keycloak launching a container - where `quarkus.oidc.client-id` is set with the value of `${oidc.client-id}` coming from the custom `ConfigSource` - as I said, `KeycloakDevServicesProcessor` checks `quarkus.oidc.client-id` when creating a default realm - without a fix provided by this PR the test fails.
 * `extensions/oidc/deployment/CodeFlowDevModeTestCase` which runs with `quarkus.oidc.auth-server-url` set - so `KeycloakDevServicesProcessor` does not do anything but `OidcDevConsoleProcessor` does prepare the properties, checks `quarkus.oidc.client-id` and again the test passes - I don't think this PR fix has any influence, it just works, also confirmed it with `quarkus-quickstarts/security-openid-connect-quickstart`

So I'm not sure how to reproduce a failure reported at #21047 without having to involve Vault - @radcortez - any ideas ?

@stuartwdouglas - re `DevConsoleRuntimeTemplateInfoBuildItem` - there are 2 cases here:
* as far as `OidcDevConsoleProcessor` is concerned and Dev UI on top of it, after updating `quarkus.oidc.client-id` in `application.properties` and refreshing Dev UI I see it is aware of the updated `quarkus.oidc.client-id` -  even though `DevConsoleRuntimeTemplateInfoBuildItem` is not used - so what can this item help with ? (note `OidcDevConsoleProcessor` runs when no `Dev Services for Keycloak` is enabled)
* if `Dev Services for Keycloak` is enabled, after updating `quarkus.oidc.client-id` in `application.properties` and refreshing Dev UI I see it is not aware of the updated `quarkus.oidc.client-id` - because when the container is not restarted - `quarkus.oidc.client-id` is not rechecked - so I guess `DevConsoleRuntimeTemplateInfoBuildItem` can help here - but restrating the application does not happen unless I refresh Dev UI - so not sure `DevConsoleRuntimeTemplateInfoBuildItem` would help much

**Comment by sberyozkin** [comment]
@stuartwdouglas - re `DevConsoleRuntimeTemplateInfoBuildItem` - there are 2 cases here:
* as far as `OidcDevConsoleProcessor` is concerned and Dev UI on top of it, after updating `quarkus.oidc.client-id` in `application.properties` and refreshing Dev UI I see it is aware of the updated `quarkus.oidc.client-id` -  even though `DevConsoleRuntimeTemplateInfoBuildItem` is not used - so what can this item help with ? (note `OidcDevConsoleProcessor` runs when no `Dev Services for Keycloak` is enabled)
* if `Dev Services for Keycloak` is enabled, after updating `quarkus.oidc.client-id` in `application.properties` and refreshing Dev UI I see it is not aware of the updated `quarkus.oidc.client-id` - because when the container is not restarted - `quarkus.oidc.client-id` is not rechecked - so I guess `DevConsoleRuntimeTemplateInfoBuildItem` can help here - but restrating the application does not happen unless I refresh Dev UI - so not sure `DevConsoleRuntimeTemplateInfoBuildItem` would help much

I'll experiment but it is probably unrelated to the issue this PR tries to address (OIDC devservices code affecting the runtime property expansion in application.properties)

**Comment by sberyozkin** [comment]
Hi @phillip-kruger @stuartwdouglas I've updated `KeycloakDevConsole` (the one aligned with `DevServices for Keycloak`) and `OidcDevConsoleProcessor` (which is used when no `DevServices for Keycloak` is started) to supply  `quarkus.oidc.client-id` (and `quarkus.oidc.credentials.secret`) to Dev UI as runtime properties and `provider.html` to refer to them as `config:property`.

When `Dev Services for Keycloak` is started and I open `Dev UI` - I get:

```
io.quarkus.qute.TemplateException: Property not found in expression {config:property('quarkus.oidc.client-id')} in template io.quarkus.quarkus-oidc/provider on line 110
	at io.quarkus.vertx.http.deployment.devmode.console.DevConsoleProcessor$6.map(DevConsoleProcessor.java:608)
```

It must be a simple bug in my PR but I can't spot what is the problem (FYI, I've also tried using `{info:'quarkus.oidc.client-id}`) which also did not work

**Comment by sberyozkin** [comment]
@stuartwdouglas @phillip-kruger Sorry for the noise, I've mistyped something in `provider.html` so after reverting it all started working.

I've also removed a ref to the unused recorder, thanks @gsmet, @stuartwdouglas for highlighting the problem - I was so certain it made a difference for making the tests work :-), but now it just works without `@Recorder`.

It should be ready to go

**Comment by sberyozkin** [comment]
I've now reverted `KeycloakDevServicesProcessor` changes completely as without `@Record` it is just 2 build steps instead of one and an extra build item, thanks

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**